### PR TITLE
Draft: The MEG [ Test rework, changing api code layout, implement ClassifyElement, implement JoinElement, Add Overseer]

### PR DESCRIPTION
**Pull request by CollinValley**
I know this is kind of a mega commit stack, I just wanted you all to take a look at it before we try to open up individual PRs.  This solves quite a few of our open issue, and I think after this is merged we can start trying to write elements, and form there create a functioning router. Which is pretty cool.  As far as I am aware, this code works as written. We could use more tests as always, but it is a good start.

### CODE REVIEW

**Comment by scgruber** [comment]
This is probably also a good opportunity to drop everything down a level so we can have multiple crates in the repo.

**Comment by CollinValley** [comment]
Moving the api to it's own crate seems like a good idea.  I can do that before pushing the official PRs.

**Comment by CollinValley** [comment]
Heads up: Resolved the Join element fairness issue this morning, commit is now in stack.

**Comment by ddowl** [comment]
We can get those first 2 commits, moving the `Element` definitions to their own modules (`086f708`, `0930c69`), into a PR straight away and get them merged. Those lgtm.

**Comment by ddowl** [comment]
I'm glad we're cleaning up the tests to be assertion-only. We should make a similar test cleanup commit for the synchronous element as well.

File: src/api/classify_element.rs [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+            }
+        }
+    }
+}
+
+impl<E: ClassifyElement> Future for ClassifyElementConsumer<E> {
+    type Item = ();
+    type Error = ();
+
+    /// Same logic as AsyncElementConsumer, except if any of the channels are full we
```
Comment (ddowl) [inline src/api/classify_element.rs:168]: We can do this in a separate PR, but if we're sharing similar logic with `AsyncElementConsumer`, we should be thinking about what parts are similar and if we can compose any `AsyncElement` components here to reduce the complexity.

File: src/api/classify_element.rs [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+                    if let Ok(task) = self.wake_providers[port].try_recv() {
+                        task.notify();
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Split Element Provider, exactly the same as AsyncElementProvider, but
```
Comment (ddowl) [inline src/api/classify_element.rs:206]: `SplitElementProvider`?

Comment (CollinValley) [reply src/api/classify_element.rs:206]: Oh needs cleanup, `ClassifyElement` used to be `SplitElement`

Comment (ddowl) [reply src/api/classify_element.rs:206]: Can we not modify `AsyncElementProvider` to have the bounds `E: Element + ClassifyElement`? I guess then the bound isn't tight enough to be useful.

File: src/api/classify_element.rs [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+            },
+            Ok(None) => {
+                Ok(Async::Ready(None))
+            },
+            Err(TryRecvError::Empty) => {
+                let task = task::current();
+                if let Err(_) = self.await_consumer.try_send(task) {
+                    task::current().notify();
+                }
+                Ok(Async::NotReady)
```
Comment (ddowl) [inline src/api/classify_element.rs:275]: This logic is repeated on line 102. We should decompose it into a function.

File: src/api/classify_element.rs (hunk 2 above)
Comment (CollinValley) [reply src/api/classify_element.rs:206]: Maybe,  I just didn't want to muck around with it at the time.  It seems possible. Though Classify doesn't do a transformation,  hence the different trait.  Classify implements `fn classify` which outputs essentially a port number with a packet.

File: src/api/classify_element.rs [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+    /// ###
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.from_consumer.try_recv() {
+            Ok(Some(packet)) => {
+                if let Ok(task) = self.wake_consumer.try_recv() {
+                        task.notify();
+                }
+                Ok(Async::Ready(Some(packet)))
+            },
+            Ok(None) => {
```
Comment (ddowl) [inline src/api/classify_element.rs:267]: It's a shame we can't use `try_ready!` here since we need the custom logic for the `Err(TryRecvError::Empty)` arm.

File: src/api/classify_element.rs [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    use crate::api::element::{Element};
+    use crate::utils::test::packet_generators::{immediate_stream, PacketIntervalGenerator};
+    use crate::utils::test::packet_collectors::ExhaustiveCollector;
+    use crossbeam::crossbeam_channel;
+    use core::time;
+
+    use futures::future::lazy;
+
+    #[allow(dead_code)]
+    struct IdentityElement {
```
Comment (ddowl) [inline src/api/classify_element.rs:296]: IdentityElement should definitely be defined elsewhere for testing purposes so we don't have to keep redefining it. We can do that in a separate commit.

File: src/api/classify_element.rs (hunk 2 above)
Comment (ddowl) [reply src/api/classify_element.rs:206]: Maybe @moosingin3space can weigh in?

File: src/api/classify_element.rs [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+            }
+        }
+    }   
+
+    #[test]
+    fn one_classify_element() {
+        let default_channel_size = 10;
+        let number_branches = 2;
+        let packets = vec![0, 1, 2, 420, 1337, 3, 4, 5, 6, 7, 8, 9, 11];
+        let packet_generator = PacketIntervalGenerator::new(time::Duration::from_millis(100), packets.clone().into_iter());
```
Comment (ddowl) [inline src/api/classify_element.rs:331]: Again for another PR, but I think we should prefer using `immediate_stream` for packet generators to keep tests speedy, and include a test for each component that confirms it can handle pauses in the input stream.

File: src/api/classify_element.rs (hunk 3 above)
Comment (CollinValley) [reply src/api/classify_element.rs:275]: Makes sense.

File: src/api/classify_element.rs (hunk 5 above)
Comment (CollinValley) [reply src/api/classify_element.rs]: Yeah I can open an issue

File: src/api/classify_element.rs (hunk 6 above)
Comment (CollinValley) [reply src/api/classify_element.rs:331]: :+1:

**Comment by ddowl** [comment]
Other than a few nitpicks I really like the `ClassifyElement`. It's implementation is _mostly_ straightforward, save for task waking issues (DAMN YOU CROSSBEAM!). The tests show a simple example, but it's descriptive and cool to see it actually work.

That being said, let's get these commits out into different PRs so we can address each of the issues in isolation.

File: src/api/classify_element.rs (hunk 1 above)
Comment (CollinValley) [reply src/api/classify_element.rs:168]: https://github.com/route-rs/route-rs/issues/29 Created issue.

File: src/api/classify_element.rs (hunk 3 above)
Comment (CollinValley) [reply src/api/classify_element.rs:275]: https://github.com/route-rs/route-rs/issues/29 Created issue

File: src/api/classify_element.rs (hunk 5 above)
Comment (CollinValley) [reply src/api/classify_element.rs:296]: https://github.com/route-rs/route-rs/issues/30

File: src/api/classify_element.rs (hunk 6 above)
Comment (CollinValley) [reply src/api/classify_element.rs:331]: https://github.com/route-rs/route-rs/issues/31

File: src/api/classify_element.rs (hunk 2 above)
Comment (CollinValley) [reply src/api/classify_element.rs:206]: https://github.com/route-rs/route-rs/issues/32

**Comment by ddowl** [comment]
Closed since we've reviewed these changes in the meetup today.
